# Lending club case study

## Introduction
Here we are going to anylyse coustomer's data from a bank and profile the coustomers weathere lending money to coustomer is a risky or a safe investment.
## Risk assement
* If the bank rejects loan to the coustomers who are likely to repay, Bank might loose business.
* If the bank lends loan to the coustomers who are not likely to repay, Bank might have financial loss.

## Discisions
* Loan rejected - No transaction with the bank any further.
* Loan accepted
    - Fully paid on time
    - Cuurent Loan has not metured yet, coustomer is paying all the payements on time.
    - Defaulted, coustomer has failed to make payment on time.

In [335]:
#Load the necessary Libraries
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline
pd.options.plotting.backend = "plotly"

In [336]:
#Read the data and explore it
loan = pd.read_csv("loan.csv")
loan.head()

/Users/rabia/opt/anaconda3/lib/python3.9/site-packages/IPython/core/interactiveshell.py:3444: DtypeWarning: Columns (47) have mixed types.Specify dtype option on import or set low_memory=False.
  exec(code_obj, self.user_global_ns, self.user_ns)


id  member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
0  1077501    1296599       5000         5000           4975.0   36 months   
1  1077430    1314167       2500         2500           2500.0   60 months   
2  1077175    1313524       2400         2400           2400.0   36 months   
3  1076863    1277178      10000        10000          10000.0   36 months   
4  1075358    1311748       3000         3000           3000.0   60 months   

  int_rate  installment grade sub_grade  ... num_tl_90g_dpd_24m  \
0   10.65%       162.87     B        B2  ...                NaN   
1   15.27%        59.83     C        C4  ...                NaN   
2   15.96%        84.33     C        C5  ...                NaN   
3   13.49%       339.31     C        C1  ...                NaN   
4   12.69%        67.79     B        B5  ...                NaN   

  num_tl_op_past_12m pct_tl_nvr_dlq  percent_bc_gt_75 pub_rec_bankruptcies  \
0                NaN            NaN               NaN                  0.0   
1                NaN            NaN               NaN                  0.0   
2                NaN            NaN               NaN                  0.0   
3                NaN            NaN               NaN                  0.0   
4                NaN            NaN               NaN                  0.0   

  tax_liens tot_hi_cred_lim total_bal_ex_mort total_bc_limit  \
0       0.0             NaN               NaN            NaN   
1       0.0             NaN               NaN            NaN   
2       0.0             NaN               NaN            NaN   
3       0.0             NaN               NaN            NaN   
4       0.0             NaN               NaN            NaN   

  total_il_high_credit_limit  
0                        NaN  
1                        NaN  
2                        NaN  
3                        NaN  
4                        NaN  

[5 rows x 111 columns]

In [337]:
# Explore the shape of the data, we have 111 columns and almost 40k entries
loan.shape

(39717, 111)

## Data Cleaning
Table is very big with lot of coloums before going any further we need to start cleaning up the data.

Lets Make sure we have unique id for every entry. We have 39717 unique Id for 39717 entry. OK !

In [338]:
loan.id.nunique()

39717

In [339]:
#Review the summary statistics for the numeric columns
loan.describe()

id     member_id     loan_amnt   funded_amnt  \
count  3.971700e+04  3.971700e+04  39717.000000  39717.000000   
mean   6.831319e+05  8.504636e+05  11219.443815  10947.713196   
std    2.106941e+05  2.656783e+05   7456.670694   7187.238670   
min    5.473400e+04  7.069900e+04    500.000000    500.000000   
25%    5.162210e+05  6.667800e+05   5500.000000   5400.000000   
50%    6.656650e+05  8.508120e+05  10000.000000   9600.000000   
75%    8.377550e+05  1.047339e+06  15000.000000  15000.000000   
max    1.077501e+06  1.314167e+06  35000.000000  35000.000000   

       funded_amnt_inv   installment    annual_inc           dti  \
count     39717.000000  39717.000000  3.971700e+04  39717.000000   
mean      10397.448868    324.561922  6.896893e+04     13.315130   
std        7128.450439    208.874874  6.379377e+04      6.678594   
min           0.000000     15.690000  4.000000e+03      0.000000   
25%        5000.000000    167.020000  4.040400e+04      8.170000   
50%        8975.000000    280.220000  5.900000e+04     13.400000   
75%       14400.000000    430.780000  8.230000e+04     18.600000   
max       35000.000000   1305.190000  6.000000e+06     29.990000   

        delinq_2yrs  inq_last_6mths  ...  num_tl_90g_dpd_24m  \
count  39717.000000    39717.000000  ...                 0.0   
mean       0.146512        0.869200  ...                 NaN   
std        0.491812        1.070219  ...                 NaN   
min        0.000000        0.000000  ...                 NaN   
25%        0.000000        0.000000  ...                 NaN   
50%        0.000000        1.000000  ...                 NaN   
75%        0.000000        1.000000  ...                 NaN   
max       11.000000        8.000000  ...                 NaN   

       num_tl_op_past_12m  pct_tl_nvr_dlq  percent_bc_gt_75  \
count                 0.0             0.0               0.0   
mean                  NaN             NaN               NaN   
std                   NaN             NaN               NaN   
min                   NaN             NaN               NaN   
25%                   NaN             NaN               NaN   
50%                   NaN             NaN               NaN   
75%                   NaN             NaN               NaN   
max                   NaN             NaN               NaN   

       pub_rec_bankruptcies  tax_liens  tot_hi_cred_lim  total_bal_ex_mort  \
count          39020.000000    39678.0              0.0                0.0   
mean               0.043260        0.0              NaN                NaN   
std                0.204324        0.0              NaN                NaN   
min                0.000000        0.0              NaN                NaN   
25%                0.000000        0.0              NaN                NaN   
50%                0.000000        0.0              NaN                NaN   
75%                0.000000        0.0              NaN                NaN   
max                2.000000        0.0              NaN                NaN   

       total_bc_limit  total_il_high_credit_limit  
count             0.0                         0.0  
mean              NaN                         NaN  
std               NaN                         NaN  
min               NaN                         NaN  
25%               NaN                         NaN  
50%               NaN                         NaN  
75%               NaN                         NaN  
max               NaN                         NaN  

[8 rows x 87 columns]

The following rows will be deleted because we do not need it in our analysis:
<br>
<br>**desc** : There is a lot of describition and we have tens of thousands different descriptions.
<br>**emp_title** : two many unique values and it is going very hard to get insights out of it (28,820 unique values).
<br>**initial_list_status** : All values equal "f", this column will not give any insight.
<br>**member_id** : We already have "id" column is unique identifier, we do not want another unique identifer.
<br>**next_pymnt_d** : We do not think the date of the next payment will have any value in this analysis.
<br>**policy_code** : All values equal "1", this column will not give any insight.
<br>**pymnt_plan** : All values equal "n", this column will not give any insight.
<br>**tax_liens** : All values equal "0" or missing value, this column will not give any insight.
<br>**url** : We do not think "url" can give any insight and we already have unique identifier column "Id".
<br>**application_type** : All values equal "INDIVIDUAL", this column will not give any insight.

In [340]:
#Drop the useless columns
coloumn_drop_list = ["desc", "emp_title", "initial_list_status", "member_id", "next_pymnt_d", "policy_code", 
                     "pymnt_plan", "tax_liens", "url", "application_type"]
loan.drop(coloumn_drop_list, inplace=True, axis=1)
loan.head()

id  loan_amnt  funded_amnt  funded_amnt_inv        term int_rate  \
0  1077501       5000         5000           4975.0   36 months   10.65%   
1  1077430       2500         2500           2500.0   60 months   15.27%   
2  1077175       2400         2400           2400.0   36 months   15.96%   
3  1076863      10000        10000          10000.0   36 months   13.49%   
4  1075358       3000         3000           3000.0   60 months   12.69%   

   installment grade sub_grade emp_length  ... num_tl_30dpd  \
0       162.87     B        B2  10+ years  ...          NaN   
1        59.83     C        C4   < 1 year  ...          NaN   
2        84.33     C        C5  10+ years  ...          NaN   
3       339.31     C        C1  10+ years  ...          NaN   
4        67.79     B        B5     1 year  ...          NaN   

   num_tl_90g_dpd_24m num_tl_op_past_12m pct_tl_nvr_dlq percent_bc_gt_75  \
0                 NaN                NaN            NaN              NaN   
1                 NaN                NaN            NaN              NaN   
2                 NaN                NaN            NaN              NaN   
3                 NaN                NaN            NaN              NaN   
4                 NaN                NaN            NaN              NaN   

  pub_rec_bankruptcies tot_hi_cred_lim total_bal_ex_mort total_bc_limit  \
0                  0.0             NaN               NaN            NaN   
1                  0.0             NaN               NaN            NaN   
2                  0.0             NaN               NaN            NaN   
3                  0.0             NaN               NaN            NaN   
4                  0.0             NaN               NaN            NaN   

   total_il_high_credit_limit  
0                         NaN  
1                         NaN  
2                         NaN  
3                         NaN  
4                         NaN  

[5 rows x 101 columns]

Filter out all the columns which all values are missing. These columns will provide no insight  since it has no values

In [341]:
#create a list of columns with all values are missing
NA_columns = list(loan.columns[loan.isna().all()])
NA_columns

['mths_since_last_major_derog',
 'annual_inc_joint',
 'dti_joint',
 'verification_status_joint',
 'tot_coll_amt',
 'tot_cur_bal',
 'open_acc_6m',
 'open_il_6m',
 'open_il_12m',
 'open_il_24m',
 'mths_since_rcnt_il',
 'total_bal_il',
 'il_util',
 'open_rv_12m',
 'open_rv_24m',
 'max_bal_bc',
 'all_util',
 'total_rev_hi_lim',
 'inq_fi',
 'total_cu_tl',
 'inq_last_12m',
 'acc_open_past_24mths',
 'avg_cur_bal',
 'bc_open_to_buy',
 'bc_util',
 'mo_sin_old_il_acct',
 'mo_sin_old_rev_tl_op',
 'mo_sin_rcnt_rev_tl_op',
 'mo_sin_rcnt_tl',
 'mort_acc',
 'mths_since_recent_bc',
 'mths_since_recent_bc_dlq',
 'mths_since_recent_inq',
 'mths_since_recent_revol_delinq',
 'num_accts_ever_120_pd',
 'num_actv_bc_tl',
 'num_actv_rev_tl',
 'num_bc_sats',
 'num_bc_tl',
 'num_il_tl',
 'num_op_rev_tl',
 'num_rev_accts',
 'num_rev_tl_bal_gt_0',
 'num_sats',
 'num_tl_120dpd_2m',
 'num_tl_30dpd',
 'num_tl_90g_dpd_24m',
 'num_tl_op_past_12m',
 'pct_tl_nvr_dlq',
 'percent_bc_gt_75',
 'tot_hi_cred_lim',
 'total_bal

In [342]:
# Delete all the NA Columns
loan.drop(NA_columns, inplace=True, axis=1)
loan.head()

id  loan_amnt  funded_amnt  funded_amnt_inv        term int_rate  \
0  1077501       5000         5000           4975.0   36 months   10.65%   
1  1077430       2500         2500           2500.0   60 months   15.27%   
2  1077175       2400         2400           2400.0   36 months   15.96%   
3  1076863      10000        10000          10000.0   36 months   13.49%   
4  1075358       3000         3000           3000.0   60 months   12.69%   

   installment grade sub_grade emp_length  ... recoveries  \
0       162.87     B        B2  10+ years  ...       0.00   
1        59.83     C        C4   < 1 year  ...     117.08   
2        84.33     C        C5  10+ years  ...       0.00   
3       339.31     C        C1  10+ years  ...       0.00   
4        67.79     B        B5     1 year  ...       0.00   

   collection_recovery_fee last_pymnt_d last_pymnt_amnt last_credit_pull_d  \
0                     0.00       Jan-15          171.62             May-16   
1                     1.11       Apr-13          119.66             Sep-13   
2                     0.00       Jun-14          649.91             May-16   
3                     0.00       Jan-15          357.48             Apr-16   
4                     0.00       May-16           67.79             May-16   

  collections_12_mths_ex_med acc_now_delinq chargeoff_within_12_mths  \
0                        0.0              0                      0.0   
1                        0.0              0                      0.0   
2                        0.0              0                      0.0   
3                        0.0              0                      0.0   
4                        0.0              0                      0.0   

  delinq_amnt  pub_rec_bankruptcies  
0           0                   0.0  
1           0                   0.0  
2           0                   0.0  
3           0                   0.0  
4           0                   0.0  

[5 rows x 47 columns]

We have identified an embeded space at the begaining of "term" column and we are deleting it using the following code

In [343]:
loan.term = loan.term.str.lstrip()

Convert the type of the columns "int_rate" and "revol_util" from string to float

In [344]:
loan.int_rate = loan.int_rate.str.rstrip('%').astype('float')
loan.revol_util = loan.revol_util.str.rstrip('%').astype('float')

Convert they type of "earliest_cr_line", "last_pymnt_d", "last_credit_pull_d" from strung to Datetime

In [345]:
loan.earliest_cr_line = pd.to_datetime(loan.earliest_cr_line, format='%b-%y')
loan.last_pymnt_d = pd.to_datetime(loan.last_pymnt_d, format='%b-%y')
loan.last_credit_pull_d = pd.to_datetime(loan.last_credit_pull_d, format='%b-%y')

Exclude any columns with all values equal to Zero

In [346]:
#Filter the data to exclude any column with all values equal to Zero
loan = loan.loc[:, (loan != 0).any(axis=0)]

In [347]:
loan.describe()

id     loan_amnt   funded_amnt  funded_amnt_inv  \
count  3.971700e+04  39717.000000  39717.000000     39717.000000   
mean   6.831319e+05  11219.443815  10947.713196     10397.448868   
std    2.106941e+05   7456.670694   7187.238670      7128.450439   
min    5.473400e+04    500.000000    500.000000         0.000000   
25%    5.162210e+05   5500.000000   5400.000000      5000.000000   
50%    6.656650e+05  10000.000000   9600.000000      8975.000000   
75%    8.377550e+05  15000.000000  15000.000000     14400.000000   
max    1.077501e+06  35000.000000  35000.000000     35000.000000   

           int_rate   installment    annual_inc           dti   delinq_2yrs  \
count  39717.000000  39717.000000  3.971700e+04  39717.000000  39717.000000   
mean      12.021177    324.561922  6.896893e+04     13.315130      0.146512   
std        3.724825    208.874874  6.379377e+04      6.678594      0.491812   
min        5.420000     15.690000  4.000000e+03      0.000000      0.000000   
25%        9.250000    167.020000  4.040400e+04      8.170000      0.000000   
50%       11.860000    280.220000  5.900000e+04     13.400000      0.000000   
75%       14.590000    430.780000  8.230000e+04     18.600000      0.000000   
max       24.590000   1305.190000  6.000000e+06     29.990000     11.000000   

       inq_last_6mths  ...  total_pymnt_inv  total_rec_prncp  total_rec_int  \
count    39717.000000  ...     39717.000000     39717.000000   39717.000000   
mean         0.869200  ...     11567.149118      9793.348813    2263.663172   
std          1.070219  ...      8942.672613      7065.522127    2608.111964   
min          0.000000  ...         0.000000         0.000000       0.000000   
25%          0.000000  ...      5112.310000      4600.000000     662.180000   
50%          1.000000  ...      9287.150000      8000.000000    1348.910000   
75%          1.000000  ...     15798.810000     13653.260000    2833.400000   
max          8.000000  ...     58563.680000     35000.020000   23563.680000   

       total_rec_late_fee    recoveries  collection_recovery_fee  \
count        39717.000000  39717.000000             39717.000000   
mean             1.363015     95.221624                12.406112   
std              7.289979    688.744771               148.671593   
min              0.000000      0.000000                 0.000000   
25%              0.000000      0.000000                 0.000000   
50%              0.000000      0.000000                 0.000000   
75%              0.000000      0.000000                 0.000000   
max            180.200000  29623.350000              7002.190000   

       last_pymnt_amnt  collections_12_mths_ex_med  chargeoff_within_12_mths  \
count     39717.000000                     39661.0                   39661.0   
mean       2678.826162                         0.0                       0.0   
std        4447.136012                         0.0                       0.0   
min           0.000000                         0.0                       0.0   
25%         218.680000                         0.0                       0.0   
50%         546.140000                         0.0                       0.0   
75%        3293.160000                         0.0                       0.0   
max       36115.200000                         0.0                       0.0   

       pub_rec_bankruptcies  
count          39020.000000  
mean               0.043260  
std                0.204324  
min                0.000000  
25%                0.000000  
50%                0.000000  
75%                0.000000  
max                2.000000  

[8 rows x 30 columns]

These 2 columns have a mix of Zeros and missing values, we are deleting them too

In [348]:
Del_columns = ['chargeoff_within_12_mths', 'collections_12_mths_ex_med']
loan.drop(Del_columns, inplace=True, axis=1)
loan.head()

id  loan_amnt  funded_amnt  funded_amnt_inv       term  int_rate  \
0  1077501       5000         5000           4975.0  36 months     10.65   
1  1077430       2500         2500           2500.0  60 months     15.27   
2  1077175       2400         2400           2400.0  36 months     15.96   
3  1076863      10000        10000          10000.0  36 months     13.49   
4  1075358       3000         3000           3000.0  60 months     12.69   

   installment grade sub_grade emp_length  ... total_pymnt_inv  \
0       162.87     B        B2  10+ years  ...         5833.84   
1        59.83     C        C4   < 1 year  ...         1008.71   
2        84.33     C        C5  10+ years  ...         3005.67   
3       339.31     C        C1  10+ years  ...        12231.89   
4        67.79     B        B5     1 year  ...         3513.33   

   total_rec_prncp total_rec_int total_rec_late_fee recoveries  \
0          5000.00        863.16               0.00       0.00   
1           456.46        435.17               0.00     117.08   
2          2400.00        605.67               0.00       0.00   
3         10000.00       2214.92              16.97       0.00   
4          2475.94       1037.39               0.00       0.00   

  collection_recovery_fee last_pymnt_d last_pymnt_amnt last_credit_pull_d  \
0                    0.00   2015-01-01          171.62         2016-05-01   
1                    1.11   2013-04-01          119.66         2013-09-01   
2                    0.00   2014-06-01          649.91         2016-05-01   
3                    0.00   2015-01-01          357.48         2016-04-01   
4                    0.00   2016-05-01           67.79         2016-05-01   

   pub_rec_bankruptcies  
0                   0.0  
1                   0.0  
2                   0.0  
3                   0.0  
4                   0.0  

[5 rows x 43 columns]

## Missing Values
<br>
<br> First we have to look for columns with any missing values

In [327]:
#explore the information about all columns in the dataframe
loan.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 39717 entries, 0 to 39716
Data columns (total 43 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   id                       39717 non-null  int64         
 1   loan_amnt                39717 non-null  int64         
 2   funded_amnt              39717 non-null  int64         
 3   funded_amnt_inv          39717 non-null  float64       
 4   term                     39717 non-null  object        
 5   int_rate                 39717 non-null  float64       
 6   installment              39717 non-null  float64       
 7   grade                    39717 non-null  object        
 8   sub_grade                39717 non-null  object        
 9   emp_length               38642 non-null  object        
 10  home_ownership           39717 non-null  object        
 11  annual_inc               39717 non-null  float64       
 12  verification_status      39717 n

In [349]:
# List of the columns with any missing values
missing_columns = loan.columns[loan.isnull().any()]
missing_columns

Index(['emp_length', 'title', 'mths_since_last_delinq',
       'mths_since_last_record', 'revol_util', 'last_pymnt_d',
       'last_credit_pull_d', 'pub_rec_bankruptcies'],
      dtype='object')

In [350]:
#Check the percentage of missing value for each column
(loan.isnull().sum() * 100 / len(loan)).sort_values(ascending=False)

mths_since_last_record     92.985372
mths_since_last_delinq     64.662487
emp_length                  2.706650
pub_rec_bankruptcies        1.754916
last_pymnt_d                0.178765
revol_util                  0.125891
title                       0.027696
last_credit_pull_d          0.005036
out_prncp                   0.000000
open_acc                    0.000000
pub_rec                     0.000000
revol_bal                   0.000000
total_acc                   0.000000
out_prncp_inv               0.000000
collection_recovery_fee     0.000000
total_pymnt                 0.000000
total_rec_prncp             0.000000
last_pymnt_amnt             0.000000
total_rec_int               0.000000
total_rec_late_fee          0.000000
recoveries                  0.000000
total_pymnt_inv             0.000000
id                          0.000000
inq_last_6mths              0.000000
loan_amnt                   0.000000
funded_amnt                 0.000000
funded_amnt_inv             0.000000
t

**mths_since_last_record** : This column has a significant high missing values, We think dropping it from the analysis will be more benificial than keeping it

In [351]:
loan.drop('mths_since_last_record', inplace=True, axis=1)

**mths_since_last_delinq** : This column has a high missing values percentage (~65%) But wew think imputing the missing values with Zero could be useful. We have noticed as the code shows beloew that all missing values are associated with zero value for "delinq_2yrs" column which mean no delinquecy occured in the last 2 years. This means it no delinq occured or it was longer than 2 years. We will imopute it with Zero but We will be careful dealing with it and come up with conclusions.

In [352]:
loan[loan['mths_since_last_delinq'].isnull()]['delinq_2yrs'].unique()

array([0])

In [353]:
loan.mths_since_last_delinq = loan.mths_since_last_delinq.fillna(0)

**emp_length
<br>pub_rec_bankruptcies
<br>last_pymnt_d
<br>revol_util
<br>title
<br>last_credit_pull_d**
<br>
<br>These columns has a low missing values percentage (0.005 - 2.7%). These columns are not good to imopute with mean or mode. It is better to remove the rows with missing values from these columns and we will lose only 2.7% of the data

In [366]:
#Removing any rows with missing values
loan = loan.dropna()

Now we have a dataframe with No missing values contains 37,825 rows

In [368]:
#information of the dataframe columns after dealing with missing values
loan.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 37825 entries, 0 to 39680
Data columns (total 42 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   id                       37825 non-null  int64         
 1   loan_amnt                37825 non-null  int64         
 2   funded_amnt              37825 non-null  int64         
 3   funded_amnt_inv          37825 non-null  float64       
 4   term                     37825 non-null  object        
 5   int_rate                 37825 non-null  float64       
 6   installment              37825 non-null  float64       
 7   grade                    37825 non-null  object        
 8   sub_grade                37825 non-null  object        
 9   emp_length               37825 non-null  object        
 10  home_ownership           37825 non-null  object        
 11  annual_inc               37825 non-null  float64       
 12  verification_status      37825 n

## Outliers

## Univariant analysis
Start to analyzie each coloum and see which are relevent. Initally steps
1. Check if the data in coloums is valid.
2. Check if there is any missing data and/or invalid data.
3. Check if there are any outliars and if outliars are needed for paticular column
4. Analyze if the coloum has any corelation with current loan status